service/matching/version_rule_helpers.go

```diff
@@ -0,0 +1,244 @@
+package matching
+
+import (
+	"fmt"
+	"slices"
+	"time"
+
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/api/taskqueue/v1"
+	"go.temporal.io/api/workflowservice/v1"
+
+	persistencepb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common"
+	hlc "go.temporal.io/server/common/clock/hybrid_logical_clock"
+	"go.temporal.io/server/common/util"
+)
+
+// checkAssignmentConditions returns an error if the new set of assignment rules don't meet the following requirements:
+// - No more rules than dynamicconfig.VersionAssignmentRuleLimitPerQueue
+// - Once assignment rules have been added, there must always be at least one "unfiltered" assignment rule, which can accept any build id
+// - No assignment rule's TargetBuildId can be a member of an existing version set
+func checkAssignmentConditions(g *persistencepb.VersioningData, maxARs int, force bool) error {
+	if force == true {
+		return nil
+	}
+	rules := slices.Clone(g.GetAssignmentRules())
+	if maxARs > 0 && len(rules) > maxARs {
+		// todo carly: should the max # of rules count rules that have been deleted but not yet cleaned up from the db?
```

Comment:
currently, I'm counting deleted (but still retained) rules in this count. The drawback of this is that if a customer really needs to add a rule urgently (ie. for a hotfix), but they have backed up rules that are waiting to be deleted, the only way for them to get around would be to do `force` or to update the max # of rules in the dynamic config. 

If we go the other way and don't count deleted rules in the max, a customer could set a ridiculously long retention period, and force the db to store an unbounded # of rules.

I guess the first way is fine since that's a good use of the force param? What are your thoughts?

Reply:
the retention period is not in the control of the user, rather in the control of server operator. The deleted rules are not exposed to the user and they don't know about the fact we keep them for some time. So the limit should be applied on non-deleted rules only. The only purpose of the retention period (at least for now) is to use it in the replication logic.